src/psyclone/psyir/nodes/intrinsic_call.py

```diff
@@ -71,51 +87,659 @@ class IntrinsicCall(Call):
+        ERFC_SCALED = IAttr(
+            'ERFC_SCALED', True, True, False,
+            ArgDesc(1, 1, DataNode), {})
+        EVENT_QUERY = IAttr(
+            'EVENT_QUERY', False, False, False,
+            ArgDesc(2, 2, DataNode), {"stat": DataNode})
+        EXECUTE_COMMAND_LINE = IAttr(
+            'EXECUTE_COMMAND_LINE', False, False, False,
+            ArgDesc(2, 2, DataNode),
+            {"wait": DataNode, "exitstat": DataNode,
+             "cmdstat": DataNode, "cmdmsg": DataNode})
+        EXP = IAttr(
+            'EXP', True, True, False,
+            ArgDesc(1, 1, DataNode), {})
+        EXPONENT = IAttr(
+            'EXPONENT', True, True, False,
+            ArgDesc(1, 1, DataNode), {})
+        EXTENDS_TYPE_OF = IAttr(
+            'EXTENDS_TYPE_OF', True, False, True,
+            ArgDesc(2, 2, DataNode), {})
+        FAILED_IMAGES = IAttr(
+            'FAILED_IMAGES', True, False, False,
+            ArgDesc(0, 0, DataNode), {"team": DataNode, "kind": DataNode})
+        FINDLOC = IAttr(
+            'FINDLOC', True, False, False,
+            ArgDesc(2, 3, DataNode),
+            {"mask": DataNode, "kind": DataNode, "back": DataNode})
+        FLOAT = IAttr(
+            'FLOAT', True, True, False,
+            ArgDesc(1, 1, DataNode), {})
+        FLOOR = IAttr(
+            'FLOOR', True, True, False,
+            ArgDesc(1, 1, DataNode), {"kind": DataNode})
+        FRACTION = IAttr(
+            'FRACTION', True, True, False,
+            ArgDesc(1, 1, DataNode), {})
+        GAMMA = IAttr(
+            'GAMMA', True, True, False,
+            ArgDesc(1, 1, DataNode), {})
+        GET_COMMAND = IAttr(
+            'GET_COMMAND', False, False, False,
+            ArgDesc(0, 0, DataNode),
+            {"command": DataNode, "length": DataNode,
+             "status": DataNode, "errmsg": DataNode})
+        GET_COMMAND_ARGUMENT = IAttr(
+            'GET_COMMAND_ARGUMENT', False, False, False,
+            ArgDesc(1, 1, DataNode),
+            {"value": DataNode, "length": DataNode,
+             "status": DataNode, "errmsg": DataNode})
+        GET_ENVIRONMENT_VARIABLE = IAttr(
+            'GET_ENVIRONMENT_VARIABLE', False, False, False,
+            ArgDesc(1, 1, DataNode),
+            {"value": DataNode, "length": DataNode, "status": DataNode,
+             "trim_name": DataNode, "errmsg": DataNode})
+        GET_TEAM = IAttr(
+            'GET_TEAM', True, False, False,
+            ArgDesc(0, 0, DataNode), {"level": DataNode})
+        HUGE = IAttr(
+            'HUGE', True, True, False,
+            ArgDesc(1, 1, (Reference, Literal)), {})
+        HYPOT = IAttr(
+            'HYPOT', True, True, False,
+            ArgDesc(2, 2, (DataNode)), {})
+        IACAHR = IAttr(
+            'IACHAR', True, True, False,
+            ArgDesc(1, 1, (DataNode)), {"kind": DataNode})
+        IALL = IAttr(
+            'IALL', True, False, False,
+            ArgDesc(1, 1, (DataNode)), {"dim": DataNode, "kind": DataNode})
```

Comment:
I don't understand the standard here. Is there any difference to it being:
(ARRAY, [DIM, MASK])
This is what I considered when giving the attributes, as DIM seems optional and not mandatory.

Reply:
I don't understand either. FWIW, Intel says:

The DIM argument to the intrinsic functions ALL, ANY, IALL, IANY, IPARITY, MAXLOC, MAXVAL, MINLOC, MINVAL, NORM2, PARITY, PRODUCT, SUM, and THIS_IMAGE may now be a present OPTIONAL dummy argument.

By which they presumably mean that it's not (or doesn't have to be) a positional argument? You'd think the standard would have it in square brackets though?

Reply:
John Reid says:
![image](https://github.com/stfc/PSyclone/assets/17409212/f6e808de-51bc-4794-b069-b2eee6db4838)

Reply:
Thanks for this @arporter  , this explains things from the Fortran perspective (they use overload instead of optional args because the return type rank is different), but I am still unsure what this means from the PSyIR.

We don't have the concept of overloading (which is not a problem because we don't do type and num args checking from call-callee). But we could also consider the combined interface of the different overloaded functions and have dim as optional (which is kind of what I did). What would you prefer?

Reply:
Coming back to this now I think we should just make `dim` optional, as in fact you did in the first place (sorry).